---
title: "Notes for the 5 accessions project"
output: word_document
---

# Biological questions for RNA-seq analysis

1. Structural variation between species
    * *Are there genes that are present/absent in particular species*?
        * **differential expression between species regardless of stage**
        * support with genomic evidence for structural variation e.g. whole-genome alignment data

2. Basic molecular mechanisms involved in branching
    * *Are there genes that might be involved in branching in all species*? These genes could encode basic mechanisms
        * **genes that are differentially expressed between PBM and SM in all species**
        * How can we support this information? *O. sativa* mutants?
    * *Are potential branching genes divergent between African and Asian rice?*
        * **comparison of genes that are differentially expressed between PBM and SM in African and Asian species**

3. Domestication and convergent evolution
    * *What are the genes that might have been under selection during the domestication of rice and convergent evolution of panicle phenotypes*?
        * **Genes where the L2FC between PBM and SM is dependent on whether or not the species is domesticated**
            * *e.g.* genes where there is a significant change between PBM and SM in *glaberrima* but not *barthi* may be involved in the African domestication
        * consider three separate domestications: Nipponbare/IR64 vs. *rufipogon* and *glaberrima* vs. *barthii*.
        * **genes implicated in at least one asian and one african domestication: convergent evolution**
        * support with genomic evidence:
            * look for variation in non-coding region of genes of interest (possible regulatory change) in these accessions
            * look for evidence of linkage disequilebrium around these genes in resequencing panels
            * possible to find transposable elements?
    * *How has domestication and selection affected known regulators of panicle architecture*?
        * Family-wise comparisons of interesting families between the 5 accessions like in the LMD paper, *e.g.* a hormone pathway, production of a metabolite, a particular TF network or family, etc.

# Other results

* Phenotyping:
    * panicle morphology (p-trap) from 5 accessions (Â± other accessions?);
    * timing of transitions in developmental stages by histology (**is this still possible with the data we have**?)
* Extended expression studies of interesting genes:
    * *e.g.* *in situ*/fluidigm/qPCR analysis of differentially expressed genes in all the developmental stages that were sampled (not just the sequenced stages).
* Hormones/metabolites?
    * Try to relate changes in hormones to changes in expression between species e.g. of genes implicated in hormone signalling
    
# Other considerations & discussion ideas

* Mapping issues:
    * Need to be clear on the question we are addressing. Aligning reads from all species against the Nipponbare genome, it will be something like:
        * Which loci, present [actually: *annotated*] in Nipponbare, have a change in expression between the primary branch meristem and spikelet meristem stage that is different between domesticated/wild, low yield/high yield, low branching/high branching species?
    * If we increase the number of allowed mismatches to get reads from non-reference species to map to the Nipponbare genome, multimapping could also increase, *e.g.* reads that previously mapped to a single location may now map to > 1. How does this balance against the increased mapping rate provided by relaxing the mismatch parameters?
        * Can try an initial mapping with normal (strict) parameters and then re-map the unmapped reads with relaxed parameters.
        * Could do an initial mapping and then call variants vs. the reference genome to create a derived transcriptome, and re-map to that. (Does this deal with indels, CNV, etc.?)
        * Gene families (or genes that share high homology with other genes in the genome) may be particularly difficult to map properly
    * Genes with more coding sequence changes may have lower mapping success. This would mean that genes under relaxed selection (*e.g.* new duplicates ) may be incorrectly given a lower expression level in non-reference species. This should be the same between stages in the same species, though.
        * "Most accurate expression information is likely to come from ... genes having a high level of expression and low rates of evolutionary change" (Hornett *et al*., 2012).
    * Genes that are consistently different between species (regardless of stage) could indicate CNV:
        * A duplication in a non-reference species could show up as increased expression in that species (if both paralogs are expressed);
        * A deletion may show up as severely reduced expression;
        * Could find these cases by doing a DEA against Nipponbare for all species using only species as a factor. (Could also control for stage?)