---
title: "Notes for the 5 accessions project"
output: word_document
---

# Results

* Phenotyping:
    * panicle morphology (p-trap) from 5 accessions (± other accessions?);
    * timing of transitions in developmental stages by histology.
* RNA seq:
    * Differential expression analysis:
        * Simple case: genes that are differentially expressed between stages in the same species;
        * To find genes that may be involved in domestication:
            * Test if the relationship between stage and gene expression is **different** between domesticated and wild species;
            * *i.e.* `expression ~ stage + domestication + (stage × domestication)`;
            * the interaction term (`stage × domestication`) tests if the effect of `stage` is **different** at different levels of the `domestication` variable.
        * Can ask other questions, *e.g.* is the relationship between stage and gene expression different in accessions with higher yield, more primary/secondary branches, etc.;
        * Find evidence to support candidates, *e.g.* selective sweeps or CNV, in genomic sequencing data.
    * Comparison between the 5 accessions of expression of genes involved in an interesting pathway, *e.g.* a hormone pathway, production of a metabolite, a particular TF network, etc.
* Extended expression studies of interesting genes:
    * *e.g.* *in situ*/fluidigm/qPCR analysis of differentially expressed genes in all the developmental stages that were sampled (not just the sequenced stages).
* Hormones/metabolites?
    * if we can relate to DE genes, or to changes in expression between species.
    
# Other considerations & discussion ideas

* Mapping issues:
    * Need to be clear on the question we are addressing. Aligning reads from all species against the Nipponbare genome, it will be something like:
        * Which loci, present [actually: *annotated*] in Nipponbare, have a change in expression between the primary branch meristem and spikelet meristem stage that is different between domesticated/wild, low yield/high yield, low branching/high branching species?
    * If we increase the number of allowed mismatches to get reads from non-reference species to map to the Nipponbare genome, multimapping could also increase, *e.g.* reads that previously mapped to a single location may now map to > 1. How does this balance against the increased mapping rate provided by relaxing the mismatch parameters?
        * Can try an initial mapping with normal (strict) parameters and then re-map the unmapped reads with relaxed parameters.
        * Could do an initial mapping and then call variants vs. the reference genome to create a derived transcriptome, and re-map to that. (Does this deal with indels, CNV, etc.?)
        * Gene families (or genes that share high homology with other genes in the genome) may be particularly difficult to map properly
    * Genes with more coding sequence changes may have lower mapping success. This would mean that genes under relaxed selection (*e.g.* new duplicates ) may be incorrectly given a lower expression level in non-reference species. This should be the same between stages in the same species, though.
        * "Most accurate expression information is likely to come from ... genes having a high level of expression and low rates of evolutionary change" (Hornett *et al*., 2012).
    * Genes that are consistently different between species (regardless of stage) could indicate CNV:
        * A duplication in a non-reference species could show up as increased expression in that species (if both paralogs are expressed);
        * A deletion may show up as severely reduced expression;
        * Could find these cases by doing a DEA against Nipponbare for all species using only species as a factor. (Could also control for stage?)