---
title: "Notes for the 5 accessions project"
output:
  word_document:
    reference_docx: /usr/share/pandoc/data/reference.docx
link-citations: yes
---

# Results
* Phenotyping
    * panicle morphology (p-trap) from 5 accessions (Â± other accessions?)
    * timing of transitions in developmental stages by histology
* RNA seq
    * Differential expression analysis
        * Simple case: genes that are DE between stages **within** each species
        * Genes that may be involved in domestication
            * the relationship between stage and gene expression is **different** between domesticated and wild species. The differential expression depends on stage AND domestication status (e.g. differentially expressed in multiple/all domesticated accessions)
        * Can ask other questions, e.g. is the relationship between stage and gene expression different in accessions with higher yield, more primary/secondary branches, etc.
        * support candidates e.g. with evidence of selective sweeps or CNV in genomic sequencing data
    * Comparison between the 5 species of expression of genes involved in an interesting pathway, e.g. a hormone pathway, production of a metabolite, a particular TF network, etc.
* Extended expression studies of interesting genes
    * e.g. *in situ*/fluidigm/qPCR analysis of differentially expressed genes in all the developmental stages that were sampled (not just the sequenced stages)
* Hormones/metabolites?
    * if we can relate to DE genes, or to changes in expression between species
    
# Discussion / other considerations
* Mapping bias
    * If we increase the number of allowed mismatches to get reads from non-reference species to map to the Nipponbare genome, multimapping could also increase, because reads that previously mapped to 1 location may now map to > 1
    * Genes with more coding sequences changes may have a lower mapping rate. This could mean that genes under relaxed selection (e.g. new duplicates ) will be given a lower expression level in non-reference species
    * Genes that are consistently different between species (regardless of stage) could indicate CNV. A duplication in a non-reference species could show up as increased expression in that species (if both paralogs are expressed); a deletion may show up as severely reduced expression. 
    